Assets/Spawner/Spawner.cs

```diff
@@ -79,6 +94,85 @@ public GameObject Spawn(EnemyData data, Waypoint startWaypoint, Vector2? pos = n
     return ObjectPool.Instance.InstantiateEnemy(data, startWaypoint, pos, parent).gameObject;
   }
 
+  // Merge an existing EnemySpawnTimes with new data derived from the other parameters.
+  // Params:
+  //   spawnTimes - Existing data about enemy spawn times to merge into.
+  //   projectedStartTime - When the current enemy will start spawning.
+  //                        If waveStartTime is nonzero, they should be equal.
+  //   spawnLocation - 0 based index of where the enemy will spawn.
+  //   repetitions - How many times the current enemy will spawn.
+  //   repeatDelay - The wait between spawn instances.
+  //   waveStartTime - When the current wave started, if it has.
+  //   enemyType - The type of enemy that will spawn.
+  public static void PopulateAndMergeSpawnTimes(
+      ref EnemySpawnTimes spawnTimes,
+      float projectedStartTime,
+      int spawnLocation,
+      int repetitions,
+      float repeatDelay,
+      float waveStartTime,
+      EnemyData.Type enemyType) {
+    // Ensure spawnTimes is big enough to reference the spawnlocation we are using.
+    while (spawnTimes.Count < (spawnLocation + 1)) {
+      spawnTimes.Add(new());
+    }
+    float remainingDuration = repetitions * repeatDelay - (Time.time - waveStartTime);
+    float endTime = projectedStartTime + remainingDuration;
+    if (!spawnTimes[spawnLocation].ContainsKey(enemyType)) {
+      spawnTimes[spawnLocation].Add(
+          enemyType,
+          new List<Tuple<float, float>> { Tuple.Create(projectedStartTime, endTime) });
+    } else {
+      List<Tuple<float, float>> enemySpawns = spawnTimes[spawnLocation][enemyType];
+      // Make sure the new interval is inserted in the correct place.
+      for (int i = 0; i < enemySpawns.Count; i++) {
+        if (projectedStartTime < enemySpawns[i].Item1) {
+          enemySpawns.Insert(i, Tuple.Create(projectedStartTime, endTime));
+          break;
+        }
+        if (enemySpawns.Count == i + 1) {
+          enemySpawns.Add(Tuple.Create(projectedStartTime, endTime));
+          break;
+        }
+      }
+      spawnTimes[spawnLocation][enemyType] = MergeEnemySpawnTimes(enemySpawns);
+    }
+  }
+
+  // Assuming spawnTimes is ordered by start time, merge any overlapping intervals and return the result.
+  private static List<Tuple<float, float>> MergeEnemySpawnTimes(List<Tuple<float, float>> spawnTimes) {
+    if (spawnTimes.Count < 2) return spawnTimes;
+
+    List<Tuple<float, float>> updatedSpawns = new();
+
+    for (int i = 0; i < spawnTimes.Count; i++) {
+      // Check for overlap.
+      if (i + 1 < spawnTimes.Count && spawnTimes[i + 1].Item1 < spawnTimes[i].Item2) {
+
+        float startTime = spawnTimes[i].Item1;
+        float endTime = Math.Max(spawnTimes[i].Item2, spawnTimes[i + 1].Item2);
+        // If i + 1 is the last element in the list, this is necessary for proper function.
+        if (spawnTimes.Count <= i + 2) i++;
```

Comment:
This is a little hard to understand, might read better if you just incriment i, and update the following logic.
Also the comment is not super helpful.

Reply:
I changed this to just naturally increment i.